src/Soap.php

```diff
@@ -1,137 +1,191 @@
 use Illuminate\Support\Traits\Macroable;
+use InvalidArgumentException;
+use RicorocksDigitalAgency\Soap\Contracts\Request;
 use RicorocksDigitalAgency\Soap\Parameters\Node;
-use RicorocksDigitalAgency\Soap\Request\Request;
+use RicorocksDigitalAgency\Soap\Response\Response;
 use RicorocksDigitalAgency\Soap\Support\Fakery\Fakery;
+use RicorocksDigitalAgency\Soap\Support\Header;
+use RicorocksDigitalAgency\Soap\Support\Scopes\HeaderSet;
+use RicorocksDigitalAgency\Soap\Support\Scopes\Inclusion;
+use RicorocksDigitalAgency\Soap\Support\Scopes\OptionSet;
+use SoapVar;
+use stdClass;
 
 /**
  * @mixin Fakery
  */
-class Soap
+final class Soap
 {
     use ForwardsCalls;
     use Macroable {
         __call as __macroableCall;
     }
 
     protected Fakery $fakery;
+
     protected Request $request;
 
-    protected $headerSets = [];
-    protected $inclusions = [];
-    protected $optionsSets = [];
-    protected $globalHooks = [];
+    /**
+     * @var array<int, HeaderSet>
+     */
+    protected array $headerSets = [];
+
+    /**
+     * @var array<int, Inclusion>
+     */
+    protected array $inclusions = [];
+
+    /**
+     * @var array<int, OptionSet>
+     */
+    protected array $optionsSets = [];
+
+    /**
+     * @var array{beforeRequesting: array<int, callable(Request): mixed>, afterRequesting: array<int, callable(Request, Response): mixed>}
+     */
+    protected array $globalHooks = [
+        'beforeRequesting' => [],
+        'afterRequesting' => [],
+    ];
 
     public function __construct(Fakery $fakery, Request $request)
     {
         $this->fakery = $fakery;
         $this->request = $request;
 
-        $this->beforeRequesting(fn ($requestInstance) => $requestInstance->fakeUsing($this->fakery->mockResponseIfAvailable($requestInstance)))
-            ->beforeRequesting(fn ($requestInstance) => $this->mergeHeadersFor($requestInstance))
-            ->beforeRequesting(fn ($requestInstance) => $this->mergeInclusionsFor($requestInstance))
-            ->beforeRequesting(fn ($requestInstance) => $this->mergeOptionsFor($requestInstance))
-            ->afterRequesting(fn ($requestInstance, $response) => $this->record($requestInstance, $response));
+        $this->beforeRequesting(fn (Request $requestInstance) => $requestInstance->fakeUsing($this->fakery->mockResponseIfAvailable($requestInstance)))
+            ->beforeRequesting(fn (Request $requestInstance) => $this->mergeHeadersFor($requestInstance))
+            ->beforeRequesting(fn (Request $requestInstance) => $this->mergeInclusionsFor($requestInstance))
+            ->beforeRequesting(fn (Request $requestInstance) => $this->mergeOptionsFor($requestInstance))
+            ->afterRequesting(fn (Request $requestInstance, Response $response) => $this->record($requestInstance, $response));
     }
 
-    public function to(string $endpoint)
+    public function to(string $endpoint): Request
     {
         return (clone $this->request)
```

Comment:
since it clones the internal request, what happens if you call Soap:to twice with different wsdls?
Does it incorrectly use the memoized client decorated client from inside the Request class?